django/contrib/admin/views/autocomplete.py

```diff
@@ -0,0 +1,92 @@
+from django.http import Http404, JsonResponse
+from django.utils.encoding import force_text
+from django.views.generic.list import BaseListView
+
+
+class AutocompleteJsonView(BaseListView):
+    """
+    View that handles requests from `.AutocompleteWidget`.
+
+    The view only supports HTTP's GET method.
+    """
+    paginate_by = 20
+    admin_site = None
+    source_model = None
+    source_model_admin = None
+
+    @property
+    def target_model(self):
+        related_field = self.source_model._meta.get_field(self.field_name)
+        return related_field.remote_field.model
+
+    @property
+    def target_model_admin(self):
+        try:
+            return self.admin_site._registry[self.target_model]
+        except KeyError:
+            raise Http404('%s is not registered in the admin.' % self.target_model)
+
+    def has_perm(self, request):
```

Comment:
Both `has_perm()` and `has_change_permission()` take an optional `obj=None` argument for object-level permission checks as implemented by e.g. django-guardian. I think we should provide this here as well.